rust/src/pgsql/pgsql.rs

```diff
@@ -0,0 +1,853 @@
+    // Just unbox...
+    std::mem::drop(unsafe { Box::from_raw(state as *mut PgsqlState) });
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_state_tx_free(state: *mut std::os::raw::c_void, tx_id: u64) {
+    let state_safe: &mut PgsqlState;
+    unsafe {
+        state_safe = cast_pointer!(state, PgsqlState);
+    }
+    state_safe.free_tx(tx_id);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_pgsql_parse_request(
+    _flow: *const Flow, state: *mut std::os::raw::c_void, pstate: *mut std::os::raw::c_void,
+    input: *const u8, input_len: u32, _data: *const std::os::raw::c_void, _flags: u8,
+) -> AppLayerResult {
+    if input_len == 0 {
+        if AppLayerParserStateIssetFlag(pstate, APP_LAYER_PARSER_EOF_TS) > 0 {
+            SCLogDebug!(" Suricata reached `eof`");
+            return AppLayerResult::ok();
+        } else {
+            return AppLayerResult::err();
+        }
+    }
+
+    let state_safe: &mut PgsqlState;
+    state_safe = cast_pointer!(state, PgsqlState);
+
+    if input == std::ptr::null_mut() && input_len > 0 {
+        // Here we have a gap signaled by the input being null, but a greater
+        // than 0 input_len which provides the size of the gap.
+        state_safe.on_request_gap(input_len);
+        AppLayerResult::ok()
+    } else {
+        let buf: &[u8];
+        buf = build_slice!(input, input_len as usize);
+        state_safe.parse_request(buf)
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_pgsql_parse_response(
+    flow: *const Flow, state: *mut std::os::raw::c_void, pstate: *mut std::os::raw::c_void,
+    input: *const u8, input_len: u32, _data: *const std::os::raw::c_void, _flags: u8,
+) -> AppLayerResult {
+    let _eof = if AppLayerParserStateIssetFlag(pstate, APP_LAYER_PARSER_EOF_TC) > 0 {
+        true
+    } else {
+        false
+    };
+    let state_safe: &mut PgsqlState;
+    state_safe = cast_pointer!(state, PgsqlState);
+
+    if input == std::ptr::null_mut() && input_len > 0 {
+        // Here we have a gap signaled by the input being null, but a greater
+        // than 0 input_len which provides the size of the gap.
+        state_safe.on_response_gap(input_len);
+        AppLayerResult::ok()
+    } else {
+        let buf: &[u8];
+        buf = build_slice!(input, input_len as usize);
+        state_safe.parse_response(buf, flow)
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_pgsql_state_get_tx(
+    state: *mut std::os::raw::c_void, tx_id: u64,
+) -> *mut std::os::raw::c_void {
+    let state_safe: &mut PgsqlState = cast_pointer!(state, PgsqlState);
+    match state_safe.get_tx(tx_id) {
+        Some(tx) => {
+            return tx as *const _ as *mut _;
+        }
+        None => {
+            return std::ptr::null_mut();
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_state_get_tx_count(state: *mut std::os::raw::c_void) -> u64 {
+    let state_safe: &mut PgsqlState;
+    unsafe {
```

Comment:
nit: is Clippy happy with these `unsafe`s inside a `fn` where raw pointers are casted?

Reply:
I may be missing something, but I think I've fixed all clippy warnings regarding unsafe code and pointers... At least, I can't see more complaints when I run `cargo +nightly clippy --allow-dirty -Z unstable-options --fix` ...

Reply:
I see. I haven't tested this branch myself but saw Jason's commit a while ago https://github.com/OISF/suricata/commit/363b5f99c36fda06f185e446ca28dff6da81fbcc which followed that rule. If it does not complain, then great!

Reply:
Ah! Clippy complained about a few places, it made me make the whole functions `unsafe`... But mostly where `build_slice!` was called, I believe...